modules/compute/src/main/java/org/apache/ignite/internal/compute/ComputeJobFailover.java

```diff
@@ -0,0 +1,158 @@
+     */
+    private final ConcurrentLinkedDeque<ClusterNode> failoverCandidates;
+
+    /**
+     * The node where the job is being executed at a given moment. If node leaves the cluster, then the job is restarted on one of the
+     * {@link #failoverCandidates} and the reference is CASed to the new node.
+     */
+    private final AtomicReference<ClusterNode> runningWorkerNode;
+
+    /** Context of the called job. Captures deployment units, jobClassName and arguments. */
+    private final RemoteExecutionContext<T> jobContext;
+
+    /**
+     * Creates a per-job instance.
+     *
+     * @param computeComponent compute component.
+     * @param nodeLeftEventsSource node left events source, used as dynamic topology service (allows to remove handlers).
+     * @param workerNode the node to execute the job on.
+     * @param failoverCandidates the set of nodes where the job can be restarted if the worker node leaves the cluster.
+     * @param units deployment units.
+     * @param jobClassName the name of the job class.
+     * @param args the arguments of the job.
+     */
+    ComputeJobFailover(
+            ComputeComponent computeComponent,
+            NodeLeftEventsSource nodeLeftEventsSource,
+            ClusterNode workerNode,
+            Set<ClusterNode> failoverCandidates,
+            List<DeploymentUnit> units,
+            String jobClassName,
+            Object... args
+    ) {
+        this.computeComponent = computeComponent;
+        this.nodeLeftEventsSource = nodeLeftEventsSource;
+        this.runningWorkerNode = new AtomicReference<>(workerNode);
+        this.failoverCandidates = new ConcurrentLinkedDeque<>(failoverCandidates);
+        this.jobContext = new RemoteExecutionContext<>(units, jobClassName, args);
+    }
+
+    /**
+     * Executes a job on the worker node and restarts the job on one of the failover candidates if the worker node leaves the cluster.
+     *
+     * @return JobExecution with the result of the job and the status of the job.
+     */
+    JobExecution<T> failSafeExecute() {
+        JobExecution<T> remoteJobExecution = launchJobOn(runningWorkerNode);
+        jobContext.initJobExecution(new FailSafeJobExecution<>(remoteJobExecution));
+
+        UUID handlerId = UUID.randomUUID();
+        nodeLeftEventsSource.addEventHandler(handlerId, new OnNodeLeft());
+        remoteJobExecution.resultAsync().whenComplete((r, e) -> nodeLeftEventsSource.removeEventHandler(handlerId));
+
+        return jobContext.failSafeJobExecution();
+    }
+
+    private JobExecution<T> launchJobOn(AtomicReference<ClusterNode> runningWorkerNode) {
+        return computeComponent.executeRemotely(runningWorkerNode.get(), jobContext.units(), jobContext.jobClassName(), jobContext.args());
+    }
+
+    class OnNodeLeft implements Consumer<ClusterNode> {
+        @Override
+        public void accept(ClusterNode leftNode) {
+            if (!runningWorkerNode.get().equals(leftNode)) {
```

Comment:
Good point. I think we should always update the candidate list when the candidate comes back after leaving the topology. This feature should be tested carefully but now it would be difficult to write such a test. I'll update the ticket https://issues.apache.org/jira/browse/IGNITE-21219